sound/soc/sof/core.c

```diff
@@ -503,11 +503,23 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 }
 EXPORT_SYMBOL(snd_sof_device_probe);
 
+static int __disable_pm(struct device *device, void *data)
+{
+	pm_runtime_disable(device);
+	return 0;
+}
+
 int snd_sof_device_remove(struct device *dev)
 {
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	struct snd_sof_pdata *pdata = sdev->pdata;
 
+	/*
+	 * Keep devices active during removal flow to avoid
```

Comment:
I don't think the comment is fully correct. if you disable the devices, they don't become 'active'. Quite the contrary, if they were suspended the framework will no longer resume them into the 'active' state.

Reply:
@plbossart Thanks, that's a very good point. I had assumed the devices are brought up, but ack, that's is just wrong. Hmm, I'll double-check, but I think I'll just update the comment and keep the code intact. The main source for delays are the state changes, we don't strictly need the devices to be active when removing.

Reply:
@plbossart @kv2019i this change seems to make sense but it makes me nervous too because it could also mess with the disable_depths for the child devices. 

If we want to do this during remove, we should also enable runtime for child devices during probe(). But of course that cant be done because some devices would need to just remain unsupported.

Reply:
I agree with @ranj063 - I think this might have unintended consequences. Instead I would look at each driver, that's performing a *real* resume during removal. Normally a driver would do that if it wants to access the hardware, right? Then we have to understand why it wants to access the hardware. If there are real valid reasons for that during driver removal, then there's no way around that and it has to be done. Otherwise we can see, if that can be avoided.
Concerning the comment, that @plbossart commented to - that means, that after `pm_runtime_disable()` when we proceed with unloading the driver, if any driver needs to access the hardware and they do `pm_runtime_get*()` or something similar, the hardware won't actually be brought up, so, the driver won't be able to perform the access? @kv2019i that doesn't sound good to me.

Reply:
@lyakh and @ranj063 Ok, thanks. I will dig into other options. The concern I have that there are awful lot of different HDA drivers. E.g. the realtek HDA codec driver gets into trouble on my ICL machine -- it takes tens of seconds to complete the resume (-> tens of seconds to rmmod the SOF driver). Many drivers have "interesting" logic to play with runtime pm to force jack detection and/or other such logic. This seems completely unnecessary when we are removing the PCI device.

You have a point though, I'll try to figure out why this one specific codec driver gets stalled, maybe that reveals something.

Reply:
Ok. Patch now updated. The root problem seems to be a genuine race. If removal starts when PCI device is runtime-suspend, device is resumed for removal but we have no syncronization to resume all devices first and then remove. Instead, we resume DSP and immediatel start unloading the topology. This leads to e.g. shutting down the DSP while HDA codec resume is still running. This can't work. 

This probably warrants more eyes looking at this (adding @dbaluta ). In V2, I now force all children to be resumed and wait until they are all powered. Then I forbid any further runtime-pm and only then proceed to unloading the machine driver.

Reply:
@kv2019i interesting. How is it running now with your patch? You don't see long delays when removing drivers any more?

Reply:
@lyakh wrote:
> @kv2019i interesting. How is it running now with your patch? You don't see long delays when removing drivers any more?

Ack. With this patch, driver module can be loaded and unloaded repeatedly without any delays. Without the patch, you will immediately hit issues.